iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/consensus/deletion/DeletionResource.java

```diff
@@ -0,0 +1,177 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.pipe.consensus.deletion;
+
+import org.apache.iotdb.commons.consensus.index.ProgressIndex;
+import org.apache.iotdb.commons.pipe.datastructure.PersistentResource;
+import org.apache.iotdb.db.pipe.event.common.schema.PipeSchemaRegionWritePlanEvent;
+import org.apache.iotdb.db.pipe.event.common.schema.PipeSchemaSerializableEventType;
+import org.apache.iotdb.db.queryengine.plan.planner.plan.node.write.DeleteDataNode;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+/**
+ * DeletionResource is designed for IoTConsensusV2 to manage the lifecycle of all deletion
+ * operations including realtime deletion and historical deletion. In order to be compatible with
+ * user pipe framework, PipeConsensus will use {@link PipeSchemaRegionWritePlanEvent}
+ */
+public class DeletionResource implements PersistentResource {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DeletionResource.class);
+  private final Consumer<DeletionResource> removeHook;
+  private final AtomicLong latestUpdateTime;
+  private PipeSchemaRegionWritePlanEvent deletionEvent;
+  private volatile Status currentStatus;
+
+  // it's safe to use volatile here to make this reference thread-safe.
+  @SuppressWarnings("squid:S3077")
+  private volatile Exception cause;
+
+  public DeletionResource(
+      PipeSchemaRegionWritePlanEvent deletionEvent, Consumer<DeletionResource> removeHook) {
+    this.deletionEvent = deletionEvent;
+    this.removeHook = removeHook;
+    this.currentStatus = Status.RUNNING;
+    latestUpdateTime = new AtomicLong(System.currentTimeMillis());
+  }
+
+  /**
+   * This method is invoked when DeletionResource is deleted by DeleteResourceManager. In this
+   * method, we release the reference of deletionEvent to resolve circular references between
+   * deletionResource and deletionEvent so that GC can reclaim them.
+   */
+  public void releaseSelf() {
+    deletionEvent = null;
+  }
+
+  public void removeSelf() {
+    removeHook.accept(this);
+  }
+
+  public void increaseReferenceCount() {
+    deletionEvent.increaseReferenceCount(DeletionResource.class.getSimpleName());
+    updateLatestUpdateTime();
+  }
+
+  public void decreaseReferenceCount() {
+    deletionEvent.decreaseReferenceCount(DeletionResource.class.getSimpleName(), false);
+  }
+
+  public long getReferenceCount() {
+    return deletionEvent.getReferenceCount();
+  }
+
+  public long getLatestUpdateTime() {
```

Comment:
What's the use of the "latestUpdateTime"?

Reply:
fixed.